### ISSUE TITLE
t.web's Request.prePathURL does not rebuild the URL correctly

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2934]
|<img alt="emgee's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| emgee reported|
|-|-|
|Trac ID|trac#2934|
|Type|defect|
|Created|2007-12-04 17:19:39Z|
|Branch|https://github.com/twisted/twisted/tree/prepath-quoting-2934|

Quick summary is that the way Request unquotes and quotes the bits of the URL is not symmetrical. This shows up when a path segment contains a '/' for example.

Request.postpath is a list of decoded, i.e. real, path segments:

```
self.postpath = map(unquote, string.split(self.path[1:], '/'))
```

Items in Request.prepath are moved to Request.prepath as child resource are located, making prepath also a list of decoded path segments. 

However, Request._prePathURL does not re-quote the individual prepath items. Instead it seems to rely on urllib.quote working (which it doesn't) for the whole URL string.

A fix can probably be taken from Nevow's flattener code. It uses the following to encode each path segment, query parameter name, query parameter value and fragment:

```
urllib.quote(original, safe="-_.!*'()")
```

<details><summary>Searchable metadata</summary>

```
trac-id__2934 2934
type__defect defect
reporter__emgee emgee
priority__highest highest
milestone__ 
branch__branches_prepath_quoting_2934 branches/prepath-quoting-2934
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1196788779000000 1196788779000000
changetime__1197295734000000 1197295734000000
version__None None
owner__ 
cc__matt@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [22049]) Branching to 'prepath-quoting-2934'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

I fixed this by quoting each segment with safe='' instead of quoting the whole URL.

This is different from what it used to do in that the hostname no longer goes through the quoting step.  I don't know if this is right or not.

Ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

I don't think that quoting hostname is a problem. It should not contain characters that needs quoting anyway.

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [22056]) Merge prepath-quoting-2934

Author: exarkun
Reviewer: therve
Fixes [#2934](https://github.com/twisted/twisted/issues/2934)

Correctly re-quote URL path segments when generating an URL string to
return from Request.prePathURL.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
